File: trunk/bee/source/libbx.c/libbx_bee_modeller.c
#include <math.h>
#include <stdlib.h>

#include "libbx_bee_modeller.h"
#include "libbx_bee_rangecoder.h"

#define BITCHAIN   4             // Size of data portion, bit
#define MAXSYMBOL 15             // Size of source alphabet, symbols
#define INCREMENT  8             // Increment of symbol frequency

/* PPM modeller's node information */

struct TNode{
  uint16_t K;               // frequency of This symbol
   uint8_t C;               // This symbol itself
   uint8_t D;               // Used FOR incoming data storage

  union {
   int32_t A;                    // source address
    struct TNode *Tear;          // Next free node
  };
  struct TNode *Next;            // Next node of This or high level
  struct TNode *Up;
};

typedef struct TNode *PNode;     // Pointer to
typedef PNode *PPNode;           // Array of nodes...

/* TBaseCoder PPM modeller struct/methods implementation */

struct TBeeModeller {
          void *Codec;
       uint32_t DictLevel;
       uint32_t Symbol;
       uint32_t Pos;
        int32_t LowestPos;       // Maximal heap Size
       uint32_t MaxCounter;      // Current heap Size
       uint32_t SafeCounter;     // Safe heap Size
       uint32_t Counter;

          PNode Heap;
         PPNode Cuts;
         PPNode List;
       uint32_t ListCount;

          PNode Root;
          PNode CurrentFreeNode;
          PNode LastFreeNode;
          PNode Tear;

       uint32_t IncreaseIndex;
       uint32_t I;
       uint32_t R;
       uint32_t Q;

          TFreq Freq;             // symbol frequencyes
     TTableCol *Part;             // Part of parameters Table
  struct TTable Table;            // parameters Table
};

PBeeModeller BeeModeller_Create(void *aCodec)
{
  PBeeModeller Self = malloc(sizeof(struct TBeeModeller));

  Self->Codec     = aCodec;
  Self->DictLevel = 0;
  Self->Freq      = malloc(sizeof(uint32_t)*(MAXSYMBOL + 1));
  Self->Heap      = NULL;
  Self->Cuts      = NULL;
  Self->List      = malloc(sizeof(uint32_t)*(MAXSYMBOL + 1));

  return Self;
}

void BeeModeller_Destroy(PBeeModeller Self)
{
  free(Self->Freq);
  free(Self->Heap);
  free(Self->Cuts);
  free(Self->List);
  free(Self);
}

static inline void BeeModeller_Add(PBeeModeller Self, uint32_t aSymbol)
{
  Self->Pos++;
  Self->LowestPos++;
  Self->Heap[Self->Pos & Self->MaxCounter].D = aSymbol;
}

static inline void BeeModeller_CreateChild(PBeeModeller Self, PNode Parent)
{
  Self->Counter++;
  PNode result = Self->CurrentFreeNode;
  if (result == Self->LastFreeNode)
  {
    result = Self->Tear;
    PNode Link = result->Tear;
    if (result->Next != NULL)
    {
      result->Next->Tear = Link;
      Link = result->Next;
    }
    if (result->Up != NULL)
    {
      result->Up->Tear = Link;
      Link = result->Up;
    }
    Self->Tear = Link;
  }
  else
    Self->CurrentFreeNode++;

  result->Next = Parent->Up;
  Parent->Up   = result;
  result->Up   = NULL;
  result->A    = Parent->A + 1;
  result->C    = Self->Heap[Parent->A & Self->MaxCounter].D;
  result->K    = INCREMENT;
}

static inline void BeeModeller_CutTail(PBeeModeller Self, PPNode I, PPNode J)
{
  PNode P = Self->Tear;
  do
  {
    (*I)->Up->Tear = P;
    P = (*I)->Up;
    (*I)->Up = NULL;
    I++;
  }
  while (!(I == J));
  Self->Tear = P;
}

static inline void BeeModeller_Cut(PBeeModeller Self)
{
  if (Self->Cuts == NULL)
  {
    Self->Cuts = malloc(sizeof(PPNode)*(Self->MaxCounter + 1));
  }

  PPNode I = &(Self->Cuts[0]);
  PPNode J = I;
  J++;

  (*I) = Self->Root;

  int32_t Bound = (Self->SafeCounter * 3) / 4;
  PNode P = NULL;
  do
  {
    P = (*I)->Up;
    do
    {
      Bound--;
      if (P->Up != NULL)
      {
        if (P->A > Self->LowestPos)
        {
          *J = P;
          J++;
        }
        else
        {
          P->Up->Tear = Self->Tear;
          Self->Tear = P->Up;
          P->Up = NULL;
        }
      }
      P = P->Next;
    }
    while (!(P == NULL));
    I++;
  }
  while (!((I == J) || (Bound < 0)));

  if (I != J)
    BeeModeller_CutTail(Self, I, J);

  Self->Counter = ((Self->SafeCounter * 3) / 4) - Bound + 1;
  Self->ListCount = 0;
}

static inline PNode BeeModeller_Tail(PBeeModeller Self, PNode Node)
{
  Node->A = Self->Pos;
  PNode result = Node->Up;

  if (result == NULL)
    BeeModeller_CreateChild(Self, Node);
  else
  {
    uint8_t C = Self->Symbol;
    if (result->C != C)
      for (;;)
      {
        PNode P = result;
        result  = result->Next;
        if (result == NULL)
        {
          BeeModeller_CreateChild(Self, Node);
          break;
        }
        else
          if (result->C == C)
          {
            P->Next = result->Next;
            result->Next = Node->Up;
            Node->Up = result;
            break;
          }
      }
  }
  return result;
}

static inline void BeeModeller_Account(PBeeModeller Self)
{
  Self->I = 0;
  Self->Q = 0;
  Self->IncreaseIndex = 0;

  uint32_t J = 0, K = 0;
  do
  {
    PNode P = Self->List[Self->I];
    if (P->Up !=NULL)
    {
      P = P->Up;
      if (Self->IncreaseIndex == 0)
        Self->IncreaseIndex = Self->I;

      if (P->Next != NULL)
      {
        // Undetermined context ...
        K = P->K * (*(Self->Part))[MAXSYMBOL + 2] >> 5;

		PNode Stored = P;
        P = P->Next;
        J = 1;
        do
        {
          J++;
          K += P->K;
          P = P->Next;
        }
        while (! (P == NULL));
        Self->Q += (*(Self->Part))[J];

        // Account:
		K = Self->R / (K + Self->Q);
		P = Stored;
		J = K * P->K * (*(Self->Part))[MAXSYMBOL + 2] >> 5;

		Self->R -= J;
        Self->Freq[P->C] += J;

        P = P->Next;
        do
        {
          J = K * P->K;
		  Self->R -= J;
          Self->Freq[P->C] += J;
		  P = P->Next;
        }
        while (!(P == NULL));
      }
      else
      {
        // Determined context ...
        K = (P->K * (*(Self->Part))[1]) / INCREMENT + 256;
		K = (Self->R / K) << 8;
        Self->Freq[P->C] += Self->R - K;
		Self->R = K;
      }
    }
    else
      if (P->A > Self->LowestPos)
      {
        // Determined context, encountered at first time ...
		BeeModeller_CreateChild(Self, P);
        K = Self->R / (*(Self->Part))[0] << 8;
        Self->Freq[P->Up->C] += Self->R - K;
        Self->R = K;
      }
    Self->I++;
  }
  while (!((Self->I == Self->ListCount) || (Self->R <= (*(Self->Part))[MAXSYMBOL + 5])));
  Self->ListCount = Self->I;
}

static inline void BeeModeller_Step(PBeeModeller Self, PRangeCod_Update Update)
{
  // ClearLongword(&Freq[0], MaxSymbol + 1);
  int32_t H;
  for (H = 0; H < MAXSYMBOL + 1; H++)
    Self->Freq[H] = 0;

  Self->R = MAXFREQ - MAXSYMBOL - 1;

  if (Self->ListCount > 0)
    BeeModeller_Account(Self);

  // Update aSymbol...
  // AddLongword(&Freq[0], MaxSymbol + 1, (R >> BitChain) + 1);
  uint32_t I = 0, J = (Self->R >> BITCHAIN) + 1;
  do
  {
    Self->Freq[I] += J;
    I++;
  }
  while (!(I == MAXSYMBOL + 1));

  Self->Symbol = Update(Self->Codec, Self->Freq, Self->Symbol);

  BeeModeller_Add(Self, Self->Symbol);

  PNode P = NULL;
  if (Self->ListCount > 0)
  {
    // Update frequencies...
    uint32_t I = 0;
    do
    {
      P = Self->List[I];
      if (I == Self->IncreaseIndex)
        P->K += INCREMENT;                // Special case...
      else
        P->K +=  (*(Self->Part))[MAXSYMBOL + 4];  // General case...
      if (P->K > (*(Self->Part))[MAXSYMBOL + 3])
        do
        {
          P->K >>= 1;
          P = P->Next;
        }
        while (!(P == NULL));
      I++;
    }
    while (!(I > Self->IncreaseIndex));

    // Update Tree...
    I = 0;
    J = I;
    do
    {
      P = BeeModeller_Tail(Self, Self->List[I]);
      if (P != NULL)
      {
        Self->List[J] = P;
        J++;
      }
      I++;
    }
    while (!(I == Self->ListCount));
    Self->ListCount = J;
  }
}

static void BeeModeller_FreshFlexible(PBeeModeller Self)
{
  Self->Tear            = NULL;
  Self->CurrentFreeNode = &(Self->Heap[0]);
  Self->LastFreeNode    = &(Self->Heap[Self->MaxCounter]);
  Self->Counter         = 0;
  Self->ListCount       = 0;
  Self->Pos             = 0;

  Self->Counter++;
  Self->Root = Self->CurrentFreeNode;
  Self->CurrentFreeNode++;

  Self->Root->Next = NULL;
  Self->Root->Up   = NULL;
  Self->Root->K    = INCREMENT;
  Self->Root->C    = 0;
  Self->Root->A    = 1;
  Self->LowestPos  = - ((int32_t) Self->MaxCounter);
}

static void BeeModeller_FreshSolid(PBeeModeller Self)
{
  if (Self->Counter > 1)
  {
    Self->ListCount = 1;
    Self->List[0]   = Self->Root;
  }
  else
    Self->ListCount = 0;
}

void BeeModeller_SetDictionary(PBeeModeller Self, uint32_t aDictLevel)
{
  Self->DictLevel   = aDictLevel;
  Self->MaxCounter  = (1 << ( 17 + Self->DictLevel)) - 1;
  Self->SafeCounter = Self->MaxCounter - 64;

  free(Self->Cuts);
  Self->Cuts = NULL;

  free(Self->Heap);
  Self->Heap = malloc(sizeof(struct TNode)*(Self->MaxCounter + 1));

  BeeModeller_FreshFlexible(Self);
}

void BeeModeller_SetTable(PBeeModeller Self, const TTableParameters *T)
{
  Self->Table.Level = (uint32_t)(*T)[0] & 0xF;

  uint32_t I = 1;
  uint32_t J , K;

  for (J = 0; J <= TABLECOLS; J++)
    for (K = 0; K <= TABLESIZE; K++)
    {
      Self->Table.T[J][K] = (int32_t)(*T)[I] + 1;
      I++;
    }

  for (I = 0; I <= 1; I++)
  {
    TTableCol *aPart;
    aPart = &(Self->Table.T[I]);

    (*aPart)[0]             =                  (*aPart)[0] + 256;               // Weight of first-encoutered deterministic symbol
    (*aPart)[MAXSYMBOL + 2] =                  (*aPart)[MAXSYMBOL + 2]  + 32;   // Recency scaling, R = R'' / 32, R'' = (R' + 1) * 32
    (*aPart)[MAXSYMBOL + 3] = INCREMENT *      (*aPart)[MAXSYMBOL + 3] << 2;    // Zero-valued parameter allowed...
    (*aPart)[MAXSYMBOL + 4] =                  (*aPart)[MAXSYMBOL + 4]  / 8;
    (*aPart)[MAXSYMBOL + 5] = floor(pow(1.082, (*aPart)[MAXSYMBOL + 5]) + 0.5); // Lowest value of interval
  }
}

static inline uint32_t BeeModeller_Update(PBeeModeller Self, uint32_t aSymbol, PRangeCod_Update Update)
{
  Self->Part = &(Self->Table.T[0]);

  uint32_t result = 0;
  Self->Symbol = aSymbol >> 0x4;
  BeeModeller_Step(Self, Update);

  result = Self->Symbol << 4;
  Self->Part = &(Self->Table.T[1]);
  Self->Symbol = aSymbol & 0xF;
  BeeModeller_Step(Self, Update);

  result += Self->Symbol;

  // Reduce Tree...
  if (Self->SafeCounter < Self->Counter)
    BeeModeller_Cut(Self);

  // Update NodeList...
  if (Self->ListCount > Self->Table.Level)
  {
    // MoveLongwordUnchecked(List[1], List[0], ListCount - 1);
    uint32_t H;
    for (H = 1; H < Self->ListCount; H++)
      Self->List[H - 1] = Self->List[H];
  }
  else
    Self->ListCount++;

  Self->List[Self->ListCount - 1] = Self->Root;
  return result;
}

inline void BeeModeller_Init(PBeeModeller Self, uint32_t DictLevel, const TTableParameters *T)
{
  BeeModeller_SetDictionary(Self, DictLevel);
  BeeModeller_SetTable(Self, T);
}

inline uint32_t BeeModeller_Encode(PBeeModeller Self, uint8_t *Buffer, uint32_t BufSize)
{
  uint32_t I;
  for (I = 0; I < BufSize; I++)
  {
    BeeModeller_Update(Self, Buffer[I], (PRangeCod_Update)BeeRangeEnc_Update);
  }
  return I;
};

inline uint32_t BeeModeller_Decode(PBeeModeller Self, uint8_t *Buffer, uint32_t BufSize)
{
  uint32_t I;
  for (I = 0; I < BufSize; I++)
  {
    Buffer[I] = BeeModeller_Update(Self, 0, (PRangeCod_Update)BeeRangeDec_Update);
  }
  return I;
};

File: bx-src/libbx_bee_modeller.pas
{
  Copyright (c) 2003-2011 Andrew Filinsky;
  Copyright (c) 2012-2014 Melchiorre Caruso.

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
}

{ Contains:

    BaseCoder class, PPM modeller;

  Modifyed:

    v0.7.8 build 0153 - 2005.07.08 by Andrew Filinsky;
    v0.7.9 build 0301 - 2007.01.23 by Andrew Filinsky;
    v0.7.9 build 0316 - 2007.02.16 by Andrew Filinsky;
  
    v1.0.0 build 2210 - 2014.01.15 by Melchiorre Caruso.
}

unit libbx_bee_modeller;

{$I bx_compiler.inc}

interface

uses
  Math,
  libbx_bee_common,
  libbx_bee_rangecoder;

const
  BitChain  = 4;                    { Size of data portion, bit                }
  MaxSymbol = 1 shl BitChain - 1;   { Size of source alphabet, symbols         }
  Increment = 8;                    { Increment of symbol frequency            }

type
  PNode  = ^TNode;                  { Pointer to modeller's node information   }
  PPNode = ^PNode;                  { Array of nodes...                        }

  { Modeller's node information }

  TNode = packed record
    Next, Up: PNode;                { Next node of this or high level          }
    K: word;                        { Frequency of this symbol                 }
    C: byte;                        { This symbol itself                       }
    D: byte;                        { Used for incoming data storage           }
    case longword of
      1: (A: longint);              { Source address                           }
      2: (Tear: PNode);             { Next free node                           }
  end;

  { PPM modeller }

  PBeeModeller = ^TBeeModeller;

  TBeeModeller = packed record
    Codec: pointer;
    DictLevel: longword;
    Symbol: longword;
    Pos: longword;
    LowestPos: longint;
    MaxCounter: longint;            { Maximal heap size                        }
    SafeCounter: longint;           { Safe heap size                           }
    Counter: longword;              { Current heap size                        }

    Heap: array of TNode;
    Cuts: array of PNode;
    List: array of PNode;
    ListCount: longword;

    Root: PNode;
    CurrentFreeNode: PNode;
    LastFreeNode: PNode;
    Tear: PNode;

    IncreaseIndex: longword;
    I: longword;
    R: longword;
    Q: longword;

    Freq:  TFreq;                   { Symbol frequencyes                       }
    Part:  ^TTableCol;              { Part of parameters Table                 }
    Table: TTable;                  { Parameters Table                         }
  end;

  function  BeeModeller_Create (aCodec: pointer): PBeeModeller;
  procedure BeeModeller_Destroy(Self: PBeeModeller);

  procedure BeeModeller_SetTableParameters(Self: PBeeModeller; Table: PByte);
  procedure BeeModeller_SetDictionaryLevel(Self: PBeeModeller; aDictLevel: longword);
  procedure BeeModeller_FreshFlexible     (Self: PBeeModeller);
  procedure BeeModeller_FreshSolid        (Self: PBeeModeller);
  function  BeeModeller_Update            (Self: PBeeModeller; aSymbol: longword; Update: PBeeRangeCodUpdate): longword;

  function  BeeModeller_Encode(Self: PBeeModeller; Buffer: PByte; BufSize: longword): longword;
  function  BeeModeller_Decode(Self: PBeeModeller; Buffer: PByte; BufSize: longword): longword;

implementation

{ TBaseCoder }

function BeeModeller_Create(aCodec: pointer): PBeeModeller;
begin
  Result :=GetMem(sizeof(TBeeModeller));

  Result^.Codec     := aCodec;
  Result^.DictLevel := 0;
  SetLength(Result^.Freq, MAXSYMBOL + 1);

  Result^.Heap := nil;
  Result^.Cuts := nil;
  SetLength(Result^.List, MAXSYMBOL + 1);
end;

procedure BeeModeller_Destroy(Self: PBeeModeller);
begin
  Self^.Freq := nil;
  Self^.Heap := nil;
  Self^.Cuts := nil;
  Self^.List := nil;
  FreeMem(Self);
end;

procedure BeeModeller_Add(Self: PBeeModeller; aSymbol: longword);
begin
  Inc(Self^.Pos);
  Inc(Self^.LowestPos);
  Self^.Heap[Self^.Pos and Self^.MaxCounter].D := aSymbol;
end;

procedure BeeModeller_CreateChild(Self: PBeeModeller; Parent: PNode);
var
  Result, Link: PNode;
begin
  Inc(Self^.Counter);
  Result := Self^.CurrentFreeNode;
  if Result = Self^.LastFreeNode then
  begin
    Result := Self^.Tear;
    Link   := Result.Tear;
    if Result.Next <> nil then
    begin
      Result.Next.Tear := Link;
      Link := Result.Next;
    end;
    if Result.Up <> nil then
    begin
      Result.Up.Tear := Link;
      Link := Result.Up;
    end;
    Self^.Tear := Link;
  end else
    Inc(Self^.CurrentFreeNode);

  Result.Next := Parent.Up;
  Parent.Up   := Result;
  Result.Up   := nil;
  Result.A    := Parent.A + 1;
  Result.C    := Self^.Heap[Parent.A and Self^.MaxCounter].D;
  Result.K    := Increment;
end;

procedure BeeModeller_Cut_Tail(Self: PBeeModeller; I, J: PPNode);
var
  P: PNode;
begin
  P := Self^.Tear;
  repeat
    I^.Up.Tear := P;
    P     := I^.Up;
    I^.Up := nil;
    Inc(I);
  until I = J;
  Self^.Tear := P;
end;

procedure BeeModeller_Cut(Self: PBeeModeller);
var
  P:     PNode;
  I, J:  PPNode;
  Bound: longint;
begin
  if Self^.Cuts = nil then
    SetLength(Self^.Cuts, Self^.MaxCounter + 1);

  I := @Self^.Cuts[0];
  J := I;
  Inc(J);

  I^    := Self^.Root;
  Bound := Self^.SafeCounter * 3 div 4;
  repeat
    P := I^.Up;
    repeat
      Dec(Bound);
      if P.Up <> nil then
        if P.A > Self^.LowestPos then
        begin
          J^ := P;
          Inc(J);
        end else
        begin
          P.Up.Tear  := Self^.Tear;
          Self^.Tear := P.Up;
          P.Up       := nil;
        end;
      P := P.Next;
    until P = nil;
    Inc(I);
  until (I = J) or (Bound < 0);

  if I <> J then
    BeeModeller_Cut_Tail(Self, I, J);

  Self^.Counter   := longint(Self^.SafeCounter * 3 div 4) - Bound + 1;
  Self^.ListCount := 0;
end;

function BeeModeller_Tail(Self: PBeeModeller; Node: PNode): PNode;
var
  P: PNode;
  C: byte;
begin
  Node.A := Self^.Pos;
  Result := Node.Up;

  if Result = nil then
    BeeModeller_CreateChild(Self, Node)
  else
  begin
    C := Self^.Symbol;
    if Result.C <> C then
      repeat
        P      := Result;
        Result := Result.Next;
        if Result = nil then
        begin
          BeeModeller_CreateChild(Self, Node);
          Break;
        end else
          if Result.C = C then
          begin
            P.Next      := Result.Next;
            Result.Next := Node.Up;
            Node.Up     := Result;
            Break;
          end;
      until False;
  end;
end;

procedure BeeModeller_Account(Self: PBeeModeller);
var
  J, K: longword;
  P, Stored: PNode;
begin
  Self^.I := 0;
  J       := 0;
  Self^.Q := 0;
  Self^.IncreaseIndex := 0;
  repeat
    P := Self^.List[Self^.I];
    if P.Up <> nil then
    begin
      P := P.Up;
      if Self^.IncreaseIndex = 0 then
        Self^.IncreaseIndex := Self^.I;

      if P.Next <> nil then
      begin
        // Undetermined context ...
        K      := P.K * Self^.Part[MaxSymbol + 2] shr 5;
        Stored := P;
        P      := P.Next;
        J      := 1;
        repeat
          Inc(J);
          Inc(K, P.K);
          P := P.Next;
        until P = nil;
        Inc(Self^.Q, Self^.Part[J]);
        // Account:
        P := Stored;
        K := Self^.R div (K + Self^.Q);
        J := K * P.K * Self^.Part[MaxSymbol + 2] shr 5;

        Dec(Self^.R, J);
        Inc(Self^.Freq[P.C], J);

        P := P.Next;
        repeat
          J := K * P.K;

          Dec(Self^.R, J);
          Inc(Self^.Freq[P.C], J);

          P := P.Next;
        until P = nil;
      end else
      begin
        // Determined context ...
        K := P.K * Self^.Part[1] div Increment + 256;
        K := (Self^.R div K) shl 8;
        Inc(Self^.Freq[P.C], Self^.R - K);
        Self^.R := K;
      end;
    end else
    if P.A > Self^.LowestPos then
    begin
      // Determined context, encountered at first time ...
      BeeModeller_CreateChild(Self, P);
      K := Self^.R div Self^.Part[0] shl 8;
      Inc(Self^.Freq[P.Up.C], Self^.R - K);
      Self^.R := K;
    end;
    Inc(Self^.I);
  until (Self^.I = Self^.ListCount) or (Self^.R <= Self^.Part[MaxSymbol + 5]);
  Self^.ListCount := Self^.I;
end;

procedure BeeModeller_Step(Self: PBeeModeller; UpdateSymbol: PBeeRangeCodUpdate);
var
  I, J: longword;
  P:    PNode;
begin
  ClearLongword(Self^.Freq[0], MaxSymbol + 1);
  Self^.R := MaxFreq - MaxSymbol - 1;

  if Self^.ListCount > 0 then
    BeeModeller_Account(Self);

  // Update aSymbol...
  AddLongword(Self^.Freq[0], MaxSymbol + 1, Self^.R shr BitChain + 1);

  Self^.Symbol := UpdateSymbol(Self^.Codec, Self^.Freq, Self^.Symbol);

  BeeModeller_Add(Self, Self^.Symbol);

  if Self^.ListCount > 0 then
  begin
    // Update frequencies...
    I := 0;
    repeat
      P := Self^.List[I];
      if I = Self^.IncreaseIndex then
        Inc(P.K, Increment)             // Special case...
      else
        Inc(P.K, Self^.Part[MaxSymbol + 4]);  // General case...
      if P.K > Self^.Part[MaxSymbol + 3] then
        repeat
          P.K := P.K shr 1;
          P   := P.Next;
        until P = nil;
      Inc(I);
    until I > Self^.IncreaseIndex;

    // Update Tree...
    I := 0;
    J := I;
    repeat
      P := BeeModeller_Tail(Self, Self^.List[I]);
      if P <> nil then
      begin
        Self^.List[J] := P;
        Inc(J);
      end;
      Inc(I);
    until I = Self^.ListCount;
    Self^.ListCount := J;
  end;
end;

procedure BeeModeller_FreshFlexible(Self: PBeeModeller);
begin
  Self^.Tear            := nil;
  Self^.CurrentFreeNode := @Self^.Heap[0];
  Self^.LastFreeNode    := @Self^.Heap[Self^.MaxCounter];
  Self^.Counter         := 0;
  Self^.ListCount       := 0;
  Self^.Pos             := 0;

  Inc(Self^.Counter);
  Self^.Root := Self^.CurrentFreeNode;
  Inc(Self^.CurrentFreeNode);

  Self^.Root.Next := nil;
  Self^.Root.Up   := nil;
  Self^.Root.K    := Increment;
  Self^.Root.C    := 0;
  Self^.Root.A    := 1;

  Self^.LowestPos := -longint(Self^.MaxCounter);
end;

procedure BeeModeller_FreshSolid(Self: PBeeModeller);
begin
  if Self^.Counter > 1 then
  begin
    Self^.ListCount := 1;
    Self^.List[0]   := Self^.Root;
  end else
    Self^.ListCount := 0;
end;

procedure BeeModeller_SetDictionaryLevel(Self: PBeeModeller; aDictLevel: longword);
begin
  Self^.DictLevel  := aDictLevel;
  Self^.MaxCounter  := (1 shl (17 + Self^.DictLevel)) - 1;
  Self^.SafeCounter := Self^.MaxCounter - 64;

  Self^.Cuts := nil;
  Self^.Heap := nil;
  SetLength(Self^.Heap, Self^.MaxCounter + 1);

  BeeModeller_FreshFlexible(Self);
end;

procedure BeeModeller_SetTableParameters(Self: PBeeModeller; Table: PByte);
var
  I, J, K:  longint;
  aPart: ^TTableCol;
begin
  Self^.Table.Level := longword(Table[0]) and $F;

  I := 1;
  for J := 0 to TABLECOLS do
    for K := 0 to TABLESIZE do
    begin
      Self^.Table.T[J][K] := longint(Table[I] + 1);
      Inc(I);
    end;

  for I := 0 to 1 do
  begin
    aPart := @Self^.Table.T[I];

    aPart[0]             :=                       aPart[0] + 256;             // Weight of first-encoutered deterministic symbol
    aPart[MaxSymbol + 2] :=                       aPart[MaxSymbol + 2]  + 32; // Recency scaling, r = r'' / 32, r'' = (r' + 1) * 32
    aPart[MaxSymbol + 3] := Increment *           aPart[MaxSymbol + 3] shl 2; // Zero-valued parameter allowed...
    aPart[MaxSymbol + 4] :=                       aPart[MaxSymbol + 4] div 8;
    aPart[MaxSymbol + 5] := Round(IntPower(1.082, aPart[MaxSymbol + 5]));     // Lowest value of interval
  end;
end;

function BeeModeller_Update(Self: PBeeModeller; aSymbol: longword; Update: PBeeRangeCodUpdate): longword;
begin
  Self^.Part   := @Self^.Table.T[0];
  Self^.Symbol := aSymbol shr $4;
  BeeModeller_Step(Self, Update);

  Result := Self^.Symbol shl 4;
  Self^.Part := @Self^.Table.T[1];
  Self^.Symbol := aSymbol and $F;
  BeeModeller_Step(Self, Update);

  Inc(Result, Self^.Symbol);

  // Reduce tree...
  if Self^.SafeCounter < Self^.Counter then
    BeeModeller_Cut(Self);

  // Update NodeList...
  if Self^.ListCount > Self^.Table.Level then
    MoveLongwordUnchecked(Self^.List[1], Self^.List[0], Self^.ListCount - 1)
  else
    Inc(Self^.ListCount);

  Self^.List[Self^.ListCount - 1] := Self^.Root;
end;

function BeeModeller_Encode(Self: PBeeModeller; Buffer: PByte; BufSize: longword): longword;
var
  I: longint;
begin
  for I := 0 to BufSize - 1 do
  begin
    BeeModeller_Update(Self, Buffer[I], @BeeRangeEnc_Update);
  end;
  Result := BufSize;
end;

function BeeModeller_Decode(Self: PBeeModeller; Buffer: PByte; BufSize: longword): longword;
var
  I: longint;
begin
  for I := 0 to BufSize - 1 do
  begin
    Buffer[I] := BeeModeller_Update(Self, 0, @BeeRangeDec_Update);
  end;
  Result := BufSize;
end;

end.

